httpclient5/src/main/java/org/apache/hc/client5/http/impl/ExecSupport.java

```diff
@@ -45,7 +45,29 @@ public static long getNextExecNumber() {
     }
 
     public static String getNextExchangeId() {
-        return String.format("ex-%010d", COUNT.incrementAndGet());
+        return createId(COUNT.incrementAndGet());
+    }
+
+    /**
+     * Create an exchange ID.
+     *
+     * Hand rolled equivalent to `String.format("ex-%010d", value)` optimized to reduce
+     * allocation and CPU overhead.
+     */
+    static String createId(long value) {
+        String longString = Long.toString(value);
+        return "ex-" + zeroPad(10 - longString.length()) + longString;
```

Comment:
Thoughts on pre-computing the ten `"ex-0..."` prefixes?

Reply:
@carterkozak how?

Reply:
Either a switch:
```java
switch (longString.length()) { // It's too bad java.lang.Long.stringSize(long) is package-private
    case 1:
        return "ex-000000000" + longString;
    /* etc... */
    case 9:
        return "ex-0" + longString;
    default:
        return "ex-" + longString;
}
```

Or an array of pre-computed values where we bounds-check and load a prefix based on the strlen of the value:
```java
private static final prefixes = new String[] {"ex-000000000", /* etc... */ ex-0"}

return len > 10  ? "ex-" + longString : prefixes[len-1] + longString;

```

Reply:
@carterkozak That would be clever.

Reply:
Testing pregenerated prefixes via https://github.com/schlosna/httpclient-format-performance/commit/5c3d850ae997b58f73e5dc9170b335fa0bd0276a doesn't actually seem worth the extra complexity

JDK 17.0.1, OpenJDK 64-Bit Server VM, 17.0.1+12-LTS
```
Benchmark                                                       Mode  Cnt     Score     Error   Units
FormatBenchmarks.stringFormat                                   avgt    5   872.268 ±  13.531   ns/op
FormatBenchmarks.stringFormat:·gc.alloc.rate                    avgt    5  1311.906 ±  20.298  MB/sec
FormatBenchmarks.stringFormat:·gc.alloc.rate.norm               avgt    5  1400.112 ±   0.010    B/op
FormatBenchmarks.stringFormat:·gc.churn.G1_Eden_Space           avgt    5  1314.251 ± 139.660  MB/sec
FormatBenchmarks.stringFormat:·gc.churn.G1_Eden_Space.norm      avgt    5  1402.548 ± 134.966    B/op
FormatBenchmarks.stringFormat:·gc.churn.G1_Survivor_Space       avgt    5     0.007 ±   0.003  MB/sec
FormatBenchmarks.stringFormat:·gc.churn.G1_Survivor_Space.norm  avgt    5     0.007 ±   0.003    B/op
FormatBenchmarks.stringFormat:·gc.count                         avgt    5    81.000            counts
FormatBenchmarks.stringFormat:·gc.time                          avgt    5    60.000                ms
FormatBenchmarks.zeroPad                                        avgt    5    42.288 ±   1.965   ns/op
FormatBenchmarks.zeroPad:·gc.alloc.rate                         avgt    5  1701.090 ±  79.519  MB/sec
FormatBenchmarks.zeroPad:·gc.alloc.rate.norm                    avgt    5    88.006 ±   0.001    B/op
FormatBenchmarks.zeroPad:·gc.churn.G1_Eden_Space                avgt    5  1709.814 ± 171.192  MB/sec
FormatBenchmarks.zeroPad:·gc.churn.G1_Eden_Space.norm           avgt    5    88.468 ±   9.798    B/op
FormatBenchmarks.zeroPad:·gc.churn.G1_Survivor_Space            avgt    5     0.009 ±   0.005  MB/sec
FormatBenchmarks.zeroPad:·gc.churn.G1_Survivor_Space.norm       avgt    5    ≈ 10⁻³              B/op
FormatBenchmarks.zeroPad:·gc.count                              avgt    5    86.000            counts
FormatBenchmarks.zeroPad:·gc.time                               avgt    5    73.000                ms
```

Reply:
The proposed implementation is far better than `String.format`, and I'm in favor of merging as is. Improvements from pre-computation will be insignificant by comparison.

That said, I think there are some small issues with the benchmark. It's using an AtomicLong counter, and incrementing for each benchmark. This helpfully matches the request ID generator, however in the benchmark we're running ~40ns operations for 45 seconds, beginning at 9,999,000,000 and incrementing roughly 1,125,000,000 times. I don't think the benchmark covers any cases where the precomputed values are used.

Reply:
Copying over @carterkozak 's updated benchmarks from https://github.com/schlosna/httpclient-format-performance/pull/1

JDK17 results
```
Benchmark                               Mode  Cnt    Score    Error  Units
FormatBenchmarks.stringFormat           avgt    5  431.935 ± 37.370  ns/op
FormatBenchmarks.zeroPadPrefixesArray   avgt    5   38.261 ±  1.599  ns/op
FormatBenchmarks.zeroPadPrefixesSwitch  avgt    5   26.841 ±  2.509  ns/op
FormatBenchmarks.zeroPadProposed        avgt    5   39.199 ±  1.526  ns/op
```


JDK8 results
```
Benchmark                               Mode  Cnt    Score    Error  Units
FormatBenchmarks.stringFormat           avgt    5  514.358 ± 61.744  ns/op
FormatBenchmarks.zeroPadPrefixesArray   avgt    5   56.109 ±  2.994  ns/op
FormatBenchmarks.zeroPadPrefixesSwitch  avgt    5   43.931 ±  6.315  ns/op
FormatBenchmarks.zeroPadProposed        avgt    5   70.111 ±  5.690  ns/op
```